include/cppcoro/io_service.hpp

```diff
@@ -178,6 +185,26 @@ namespace cppcoro
 
 	};
 
+	template<typename TASK>
+	decltype(auto) io_service::process_events_until_complete(TASK&& task)
```

Comment:
This implementation should work just fine for the cases where there is only a single thread processing events, but if there are multiple threads processing events then calling `stop()` when the task completes can potentially lead those other threads to exit their event loops too (whether or not they do actually exit is racy since the waiting thread calls `reset()` without necessarily waiting for all other threads to exit their loops). Ideally only the current thread would exit the event loop in this case.

To implement support for this behaviour would require integration into the core event loop within `io_service` to allow it to check the status of the task/awaitable after processing each IOCP event. It would also likely need to use a short timeout on the call to `GetQueuedCompletionStatus()` to allow it to poll for the task/awaitable potentially completing on another thread.

It will also need to change a little once the `generic_ops` branch is merged as the `[shared_]task<T>::get_starter()` methods and `cppcoro::detail::continuation` type have both been removed in that work. I'll update the implementation as appropriate when merging this work into the `generic_ops` branch.

Reply:
To be clear, I'm happy with the implementation as it is for now.

I'll think further about whether it makes sense to limit its use to the single-threaded event processing case or to use a more complicated implementation that allows for boost-blocking use-cases with multiple I/O threads.

Reply:
I'll try to make it work with multiple threads. I think it may be possible to do it without polling. FWIW I think that this function is semantically similar to `sync_wait`, so if one is not limited to a single thread, another shouldn't be as well.